src/fraction/approx/sqrt.rs

```diff
@@ -0,0 +1,637 @@
+                panic!("cannot take the square root of a negative number: {}", self)
+            }
+
+            // Short-circuit for zero so we don't have to worry about it when checking the signs.
+            GenericFraction::Rational(_, r) if r.is_zero() => RawApprox::Zero,
+
+            GenericFraction::Rational(sign, ratio)
+                if {
+                    // `num` doesn't seem to give us a way to check the sign of a `T` given the
+                    // current trait bounds we have on it, but we can work around this by checking
+                    // against zero.
+
+                    let numer_sign = if ratio.numer() > &T::zero() {
+                        Sign::Plus
+                    } else {
+                        Sign::Minus
+                    };
+
+                    // Combine the whole fraction's sign with the sign of the numerator. This
+                    // allows us to recognise stuff like -(-1/2) as being positive, for example.
+                    let numer_sign = *sign * numer_sign;
+
+                    let denom_sign = if ratio.denom() > &T::zero() {
+                        Sign::Plus
+                    } else {
+                        Sign::Minus
+                    };
+
+                    numer_sign != denom_sign
+                } =>
+            {
+                panic!("cannot take the square root of a negative number: {}", self)
+            }
+
+            _positive => self.sqrt_abs_with_accuracy_raw(accuracy),
+        }
+    }
+
+    /// Returns a rational approximation of the principal square root of `self`. `accuracy` refers
+    /// to the square of the return value relative to `self`.
+    ///
+    /// See the [module-level documentation](`self`) for more details.
+    ///
+    /// # Panics
+    /// This method will panic if `self` is negative.
+    pub fn sqrt_with_accuracy(&self, accuracy: impl Borrow<Accuracy>) -> GenericFraction<BigUint> {
+        self.sqrt_with_accuracy_raw(accuracy).simplified().into()
+    }
+
+    /// Returns an unsimplified rational approximation of the principal square root of `self`.
+    /// `decimal_places` refers to the accuracy of the square of the return value relative to
+    /// `self`.
+    ///
+    /// See the [module-level documentation](`self`) for more details.
+    ///
+    /// # Panics
+    /// This method will panic if `self` is negative.
+    pub fn sqrt_raw(&self, decimal_places: u32) -> RawApprox {
+        self.sqrt_with_accuracy_raw(Accuracy::decimal_places(decimal_places))
+    }
+
+    /// Returns a rational approximation of the principal square root of `self`. `decimal_places`
+    /// refers to the accuracy of the square of the return value relative to `self`.
+    ///
+    /// See the [module-level documentation](`self`) for more details.
+    ///
+    /// # Panics
+    /// This method will panic if `self` is negative.
+    pub fn sqrt(&self, decimal_places: u32) -> GenericFraction<BigUint> {
```

Comment:
I feel like `panic` is an overkill here.

Imagining this code used in external applications where people would use libraries using other libraries. The end application might inherently invoke this code with negative user input, which would lead to the entire application going down. Also, given a lot of anecdotes about _security vulnerabilities_ in the applications caused by people being lazy to use longer method names (thus, leaning towards `sqrt` instead of `sqrt_abs`, which also feels like the _default_ method).

I reckon it might be safer to replicate the default Rust `f64` behaviour (and adhere to the Float trait contract) - https://docs.rs/num/latest/num/trait.Float.html#tymethod.sqrt

That would mean, we could return `NaN` instead of triggering the panic in here.

What do you think?

Reply:
Good point here – I hadn't considered comparing to the `f64` API. I guess there's no reason not to return `NaN`, since `GenericFraction` has the ability to represent it. I'll get that fixed. Cheers.